src/combinatorics/composition.lean

```diff
@@ -0,0 +1,680 @@
+      { exact h.2 },
+      { exact h.1 } },
+    { rw fin.ext_iff,
+      exact nat.add_sub_cancel' h.1 } }
+end
+
+/-- The embeddings of different blocks of a composition are disjoint. -/
+lemma disjoint_range {i₁ i₂ : fin c.length} (h : i₁ ≠ i₂) :
+  disjoint (set.range (c.embedding i₁)) (set.range (c.embedding i₂)) :=
+begin
+  classical,
+  wlog h' : i₁ ≤ i₂ using i₁ i₂,
+  { by_contradiction d,
+    obtain ⟨x, hx₁, hx₂⟩ :
+      ∃ x : fin n, (x ∈ set.range (c.embedding i₁) ∧ x ∈ set.range (c.embedding i₂)) :=
+    set.not_disjoint_iff.1 d,
+    have : i₁ < i₂ := lt_of_le_of_ne h' h,
+    have A : (i₁ : ℕ).succ ≤ i₂ := nat.succ_le_of_lt this,
+    apply lt_irrefl (x : ℕ),
+    calc (x : ℕ) < c.size_up_to (i₁ : ℕ).succ : (c.mem_range_embedding_iff.1 hx₁).2
+    ... ≤ c.size_up_to (i₂ : ℕ) : monotone_sum_take _ A
+    ... ≤ x : (c.mem_range_embedding_iff.1 hx₂).1 },
+  { rw disjoint.comm,
+    apply this (ne.symm h) }
+end
+
+lemma mem_range_embedding (j : fin n) :
+  j ∈ set.range (c.embedding (c.index j)) :=
+begin
+  have : c.embedding (c.index j) (c.inv_embedding j)
+    ∈ set.range (c.embedding (c.index j)) := set.mem_range_self _,
+  rwa ← c.embedding_comp_inv j at this
+end
+
+lemma mem_range_embedding_iff' {j : fin n} {i : fin c.length} :
+  j ∈ set.range (c.embedding i) ↔ i = c.index j :=
+begin
+  split,
+  { rw ← not_imp_not,
+    assume h,
+    exact set.disjoint_right.1 (c.disjoint_range h) (c.mem_range_embedding j) },
+  { assume h,
+    rw h,
+    exact c.mem_range_embedding j }
+end
+
+/-- Two compositions (possible of different integers) coincide if and only if they have the
+same sequence of blocks of positive integers. -/
+lemma sigma_eq_iff_blocks_pnat_eq {c : Σ n, composition n} {c' : Σ n, composition n} :
+  c = c' ↔ c.2.blocks_pnat = c'.2.blocks_pnat :=
+begin
+  refine ⟨λ H, by rw H, λ H, _⟩,
+  rcases c with ⟨n, c⟩,
+  rcases c' with ⟨n', c'⟩,
+  have : n = n', by { rw [← c.blocks_pnat_sum, ← c'.blocks_pnat_sum, H] },
+  induction this,
+  simp only [true_and, eq_self_iff_true, heq_iff_eq],
+  ext1,
+  exact H
+end
+
+/-- Two compositions (possible of different integers) coincide if and only if they have the
+same sequence of blocks. -/
+lemma sigma_eq_iff_blocks_eq {c : Σ n, composition n} {c' : Σ n, composition n} :
+  c = c' ↔ c.2.blocks = c'.2.blocks :=
+begin
+  refine ⟨λ H, by rw H, λ H, _⟩,
+  rwa [sigma_eq_iff_blocks_pnat_eq, ← (injective_map_iff.2 subtype.val_injective).eq_iff]
+end
+
+end composition
+
+
+/-!
+### Compositions as sets
+
+Combinatorial viewpoints on compositions, seen as finite subsets of `fin (n+1)` containing `0` and
+`n`, where the points of the set (but `n`) correspond to the leftmost points of each block.
+-/
+
+/-- Bijection between compositions of `n` and subsets of `{0, ..., n-2}`, defined by
+considering the restriction of the subset to `{1, ..., n-1}` and shifting to the left by one. -/
+def composition_as_set_equiv (n : ℕ) : composition_as_set n ≃ finset (fin (n - 1)) :=
```

Comment:
This one is nasty! I thought "surely it can't be this hard" ... but ... didn't get very far.